renter/rpc.go

```diff
@@ -308,46 +308,57 @@ func (s *Session) FormContract(renterKey types.PrivateKey, renterFunds, hostColl
 		hostInputs[i].Signatures = hostSigs.SiacoinInputSignatures[i]
 	}
 
+	// the transaction should now be valid
+	if err := vc.ValidateTransaction(txn); err != nil {
```

Comment:
Hmm. This is tricky. If the renter and host aren't on the same block, then their input proofs will be out-of-sync, and the transaction will be invalid. Worse, we can't simply lean on the txpool's "flexibility" in updating proofs, because the txpool will call `UpdateElementProof` on *all* of the elements. So to make this work, we need the renter and host to agree on a validation context to use.

This feels like it will be pretty ugly to deal with properly. On the bright side, the Utreexo proofs aren't actually *creating* a new problem -- they're revealing a problem that was always there. Any consensus mismatch between the renter and host creates the potential for "races" like this that can result in invalid transactions.

Reply:
So we'll need to add some form of synchronization to the RHP. Maybe the best API for this would be for the renter and host to send their `vc.Index` alongside their inputs. Then, we can add a generic proof-updating function to `ChainManager`:

```go
func (cm *ChainManager) UpdateElementProof(e *types.StateElement, cur, dst types.ChainIndex) error
```

Reply:
I went ahead and implemented this in https://github.com/SiaFoundation/core/commit/40ae1d57010c5ac814d72ea8c988161953918c97

IMO we should just merge this PR as-is so that #69 can proceed, and tackle the sync issue in a follow up.

Reply:
Any value we send, we have to deal with a small race between the call to `cm.TipContext` and `w.FundTransaction`... The easiest fix would be to send the renter's `ChainIndex` in the initial `ContractRequest` and make the host slightly flexible, like the txpool. Created #85 for discussion and LGTM.